### PR TITLE
add kernel flag to override ntp conf

### PR DESCRIPTION
**Pull request by Omarabdul3ziz**
### Description
ntp flag is a comma seperated list of ntp servers, which will override the content in /etc/ntp.conf

### Changes
ntp flag is a comma seperated list of ntp servers, which will override the content in /etc/ntp.conf

### Related Issues
- https://github.com/threefoldtech/zos/issues/2219

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstring

### CODE REVIEW

File: cmds/internet/main.go [hunk 1]
```diff
@@ -215,6 +219,19 @@ func configureZOS() error {
...
+			servers := strings.Split(*env.NTPServer, ",")
+			var buffer strings.Builder
+			for _, server := range servers {
+				buffer.WriteString(fmt.Sprintf("server %s\n", server))
+			}
+			err := os.WriteFile(ntpFilePath, []byte(buffer.String()), 0644)
+			if err != nil {
+				return errors.Wrap(err, "failed to write ntp server conf")
+			}
+		}
```
Comment (xmonader) [inline cmds/internet/main.go]: Always check with strings.TrimSpace(var) != ""

Comment (xmonader) [inline cmds/internet/main.go]: these can be joined  `if err:= ... ; err != nil`

File: cmds/internet/main.go [hunk 2]
```diff
@@ -215,6 +219,19 @@ func configureZOS() error {
 			}
 		}
 
+		if env.NTPServer != nil && *env.NTPServer != "" {
+			// override /etc/ntp.conf
+			servers := strings.Split(*env.NTPServer, ",")
+			var buffer strings.Builder
+			for _, server := range servers {
```
Comment (xmonader) [inline cmds/internet/main.go:226]: No validation on the server variable before generating a broken ntp config file?

File: pkg/environment/environment.go [hunk 3]
```diff
@@ -250,6 +253,12 @@ func getEnvironmentFromParams(params kernel.Params) (Environment, error) {
 		}
 	}
 
+	if ntps, ok := params.Get("ntp"); ok {
```
Comment (xmonader) [inline pkg/environment/environment.go]: Should we use GetOne?

File: qemu/README.md [hunk 4]
```diff
@@ -115,9 +115,12 @@ sudo dnsmasq --strict-order \
...
 
-where `runmode` is one of `dev` , `test`  or `prod`, 
-      `farmer_id` is the id of the farm you registered with `tffarmer`
-      `ssh-user` is github username provided if a user need to pass ssh-key to the node
+### Flags and kernel args:
+
+- `runmode` is one of `dev` , `test`  or `prod`.
+- `farmer_id` is the id of the farm you registered with `tffarmer`
+- `ssh-user` is github username provided if a user need to pass ssh-key to the node
+- `ntp` a comma separated list with ntp servers to override `/etc/ntp.conf`
```
Comment (xmonader) [inline qemu/README.md:123]: Add an example

File: cmds/internet/main.go [hunk 5]
```diff
@@ -215,6 +219,19 @@ func configureZOS() error {
...
+				buffer.WriteString(fmt.Sprintf("server %s\n", server))
```
Comment (xmonader) [inline cmds/internet/main.go]: The server should have space trimmed, but it won't matter given the whitespace tolerance in the ntp file

**Review by xmonader — CHANGES_REQUESTED** [review]

File: cmds/internet/main.go (hunk 2 above)
Comment (Omarabdul3ziz) [reply cmds/internet/main.go:226]: hmm other than not empty i think we can also validate if the server is suffixed with `.pool.ntp.org`?

**Review by muhamadazmy — CHANGES_REQUESTED** [review]
Thanks @Omarabdul3ziz for your work. But (may be i should have said it earlier) I think the changes related to NTP should be part of 0-initramfs.

- First of all, ntp is not managed by zos repo, but it's part of 0-initramfs, it's fully installed and configured by this repo.
- The change definitely does not belong to `internet`. If you had doubts where you put the code you should have checked with me. Why internet should care about what ntp config is, right ?
- I noticed you don't even restart the service after updating it's config. Why do you think config changes will take effect. Note that, ntp is already running at this stage. If you check the ntp.yaml you see that it does not has any dependencies, hence it will get started (probably just give few errors until the node has internet access before it resumes)

Taking all of this into account what I think what should be done is the following:
- Changes has to be part of 0-initramfs 
- The changes are pretty simple, just modify the ntp.yaml file to do the following:
  - this has to be done in bash (more specifically `sh`)  
  - check if `ntp=` is passed as a cmdline argument `/proc/cmdline`
  - if set, create the file, with proper config (no validation since it's on the user now)
  - `exec ntpd -n` as last line of the script 

this will make sure that ntp is definitely configured before starting

**Pull request closed without merging**